File: template_method/java/Gamer.java
abstract class Game {

	abstract void initialize();

	abstract void startPlay();

	abstract void endPlay();

	// main method
	final void play() {
		initialize();
		startPlay();
		endPlay();
	}
}

final class Mario extends Game {

	@Override
	public void initialize() {
		System.out.println("Mario Game Initialized! Start playing.");
	}

	@Override
	public void startPlay() {
		System.out.println("Mario Game Started. Enjoy the game!");
	}

	@Override
	public void endPlay() {
		System.out.println("Mario Game Finished!");
	}
}

final class Sonic extends Game {

	@Override
	public void initialize() {
		System.out.println("Sonic Game Initialized! Start playing.");
	}

	@Override
	public void startPlay() {
		System.out.println("Sonic Game Started. Enjoy the game!");
	}

	@Override
	public void endPlay() {
		System.out.println("Sonic Game Finished!");
	}
}

final class Gamer {

	public static void main(String[] args) {
		Game game = new Mario();
		game.play();

		game = new Sonic();
		game.play();
	}
}

File: template_method/kotlin/gamer.kt
abstract class Game {

	abstract fun initialize(): Unit;

	abstract fun startPlay(): Unit;

	abstract fun endPlay(): Unit;

	// main method
	fun play(): Unit {
		initialize();
		startPlay();
		endPlay();
	}
}

class Mario : Game() {

	override fun initialize(): Unit {
		System.out.println("Mario Game Initialized! Start playing.");
	}

	override fun startPlay(): Unit {
		System.out.println("Mario Game Started. Enjoy the game!");
	}

	override fun endPlay(): Unit {
		System.out.println("Mario Game Finished!");
	}
}

class Sonic : Game() {

	override fun initialize(): Unit {
		System.out.println("Sonic Game Initialized! Start playing.");
	}

	override fun startPlay(): Unit {
		System.out.println("Sonic Game Started. Enjoy the game!");
	}

	override fun endPlay(): Unit {
		System.out.println("Sonic Game Finished!");
	}
}

fun main(): Unit {
	var gamer: Game = Mario()
	gamer.play()

	gamer = Sonic()
	gamer.play()
}